### PR TITLE
daemon: make /v2/download take snapRevisionOptions

### PR DESCRIPTION
**Pull request by mvo5**
The current /v2/download endpoint only supports downloading
by name. No options like "channel", "revision" or "cohort" are
supported. We will need those options for the "snap download"
command that goes through the snapd API.

This PR implements passing those options. It also contains two
drive-by fixes in store.go. One typo fix (thanks to John) and an
update from the click package index to the snapstore API docs.

### CODE REVIEW

File: daemon/api_download.go [hunk 1]
```diff
@@ -62,20 +63,39 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
...
+		InstanceName: snapName,
+		Revision:     dlOpts.Revision,
+		CohortKey:    dlOpts.CohortKey,
+		Channel:      dlOpts.Channel,
+	}}
+	snaps, err := getStore(c).SnapAction(context.TODO(), nil, actions, user, nil)
 	if err != nil {
 		return SnapNotFound(snapName, err)
 	}
+	info := snaps[0]
```
Comment (anonymouse64) [inline daemon/api_download.go]: should we also have a check here that we didn't get more than one snap returned somehow?

File: store/store.go [hunk 2]
```diff
@@ -1642,9 +1642,9 @@ func (s *Store) DownloadStream(ctx context.Context, name string, downloadInfo *s
 	return resp.Body, nil
 }
 
-var doDownloadReq = doDowloadReqImpl
+var doDownloadReq = doDownloadReqImpl
```
Comment (anonymouse64) [inline store/store.go]: nice catch

File: store/store.go [hunk 3]
```diff
@@ -386,7 +386,7 @@ func New(cfg *Config, dauthCtx DeviceAndAuthContext) *Store {
 
 // API endpoint paths
 const (
-	// see https://wiki.ubuntu.com/AppStore/Interfaces/ClickPackageIndex
+	// see https://dashboard.snapcraft.io/docs/
```
Comment (anonymouse64) [inline store/store.go]: :+1:

File: daemon/api.go [hunk 4]
```diff
@@ -731,20 +731,25 @@ func (*licenseData) Error() string {
 	return "license agreement required"
 }
 
+type snapDownloadOptions struct {
```
Comment (pedronis) [inline daemon/api.go]: don't think Download is a great name for this set of options.  snapstate has something very similar but internal called RevisionOptions, I would follow the lead from there and call these snapRevisionOptions, include LeaveCohort as well.

File: daemon/api.go [hunk 5]
```diff
@@ -731,20 +731,25 @@ func (*licenseData) Error() string {
...
-	Revision         snap.Revision `json:"revision"`
-	CohortKey        string        `json:"cohort-key"`
-	LeaveCohort      bool          `json:"leave-cohort"`
-	DevMode          bool          `json:"devmode"`
-	JailMode         bool          `json:"jailmode"`
-	Classic          bool          `json:"classic"`
-	IgnoreValidation bool          `json:"ignore-validation"`
-	Unaliased        bool          `json:"unaliased"`
-	Purge            bool          `json:"purge,omitempty"`
+	snapDownloadOptions
```
Comment (pedronis) [inline daemon/api.go]: Then place those just where Revision etc was in the struct.  It's strange for this not to lead with Action.

File: daemon/api.go (hunk 4 above)
Comment (mvo5) [reply daemon/api.go]: Thanks, fixed.

File: daemon/api.go (hunk 5 above)
Comment (mvo5) [reply daemon/api.go]: Indeed, thanks! Fixed as well.

File: daemon/api_download.go [hunk 6]
```diff
@@ -63,20 +64,39 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
 		return BadRequest("download operation requires action")
 	}
 
+	var opts *snapRevisionOptions
+	if len(action.Options) == 1 {
+		opts = &action.Options[0]
+	}
+	if len(action.Options) > 1 {
```
Comment (bboozzoo) [inline daemon/api_download.go]: Nitpick: I'd replace the order of checks, first bail out on len(action.Options) > 1 and then len(action.Options) == 1

File: daemon/api_download.go [hunk 7]
```diff
@@ -37,8 +37,9 @@ var snapDownloadCmd = &Command{
 
 // SnapDownloadAction is used to request a snap download
 type snapDownloadAction struct {
-	Action string   `json:"action"`
-	Snaps  []string `json:"snaps,omitempty"`
+	Action  string                `json:"action"`
+	Snaps   []string              `json:"snaps,omitempty"`
```
Comment (pedronis) [inline daemon/api_download.go]: not new in this PR but why does this have multiple snaps and action at all?

Comment (chipaca) [reply daemon/api_download.go]: there was no strong reason; I don't particularly like it either. I was close to flagging it in the original PR but it is the more generic solution so I could see where the author was coming from.

I'm ok with dropping them in the name of YAGNI.

Comment (mvo5) [reply daemon/api_download.go]: This is now updated based on https://github.com/snapcore/snapd/pull/7560

File: daemon/api_download.go [hunk 8]
```diff
@@ -37,7 +37,8 @@ var snapDownloadCmd = &Command{
 
 // SnapDownloadAction is used to request a snap download
 type snapDownloadAction struct {
-	SnapName string `json:"snap-name,omitempty"`
+	SnapName string              `json:"snap-name,omitempty"`
+	Options  snapRevisionOptions `json:"options,omitempty"`
```
Comment (pedronis) [inline daemon/api_download.go]: this probably should be embedded, we don't have "options" in any other API

File: daemon/api_download.go (hunk 1 above)
Comment (mvo5) [reply daemon/api_download.go]: This is now checked.

File: daemon/api_download.go (hunk 8 above)
Comment (mvo5) [reply daemon/api_download.go]: Thanks! I embedded this now and refactored/updated the tests a bit.

File: daemon/api_download.go [hunk 9]
```diff
@@ -54,23 +55,34 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
...
+		Action:       "download",
+		InstanceName: opts.SnapName,
+		Revision:     opts.Revision,
+		CohortKey:    opts.CohortKey,
+		Channel:      opts.Channel,
+	}}
+	snaps, err := getStore(c).SnapAction(context.TODO(), nil, actions, user, nil)
 	if err != nil {
-		return SnapNotFound(snapName, err)
+		return SnapNotFound(opts.SnapName, err)
```
Comment (pedronis) [inline daemon/api_download.go]: we probably want to use:

errToResponse(err, []string{action.snapName}, InternalError, "cannot download: %v")

here

**Pull request merged**